### PR TITLE
Feature/tokenization

### PR DESCRIPTION
**Pull request by rrahn**
Initial version.
Please comment and do not merge yet.

### CODE REVIEW

File: include/seqan3/core/meta/associated_types.hpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+iterator_type_helper(t *);
+
+template <typename t>
+ranges::v3::detail::object_remove_cv<typename t::const_iterator>
+iterator_type_helper(t const *);
+
+template <typename t>
+using iterator_type_ = meta::_t<decltype(detail::iterator_type_helper(ranges::v3::nullptr_v<t>))>;
+}  // namespace seqan3::detail
+
```
Comment (h-2) [inline include/seqan3/core/meta/associated_types.hpp]: this is what we have std::iterator_traits for!

File: include/seqan3/core/meta/associated_types.hpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+using size_type_t = typename size_type<t>::type;
+
+// TODO(rrahn): iterator_category
+
+template <typename t>
+struct iterator_type : meta::defer<detail::iterator_type_, t>
+{};
+
+template <typename t>
+using iterator_type_t = typename iterator_type<t>::type;
```
Comment (h-2) [inline include/seqan3/core/meta/associated_types.hpp]: I would propose just using the ranges metafunctions. If you really think we need them in our namespaces, please don't forward the metafunction, just "use" it, e.g:
```cpp
using ranges::value_type_t;
```

File: include/seqan3/io/concept/stream.hpp [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+ template <typename char_t, typename traits_t>
+ struct stream_traits<std::basic_ostringstream<char_t, traits_t>>
+ {
+     using stream_ = std::basic_ostringstream<char_t, traits_t>;
+     using char_type   = typename stream_::char_type;
+     using traits_type = typename stream_::traits_type;
+     using int_type    = typename stream_::int_type;
+     using off_type    = typename stream_::off_type;
+     using pos_type    = typename stream_::pos_type;
+ };
```
Comment (h-2) [inline include/seqan3/io/concept/stream.hpp]: lots of code copy'n'paste, you could use one struct template and constrain it to the list of types thereby avoiding copy'n'paste.

File: include/seqan3/io/concept/stream.hpp [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+{
+    typename char_type<t>::type;
+    typename traits_type<t>::type;
+    typename int_type<t>::type;
+    typename pos_type<t>::type;
+    typename off_type<t>::type;
+
+    // formatted output.
+    { operator<<(stream, typename t::char_type{}) } -> auto&;
+};
```
Comment (h-2) [inline include/seqan3/io/concept/stream.hpp]: First you define structs with members, than you define free metafunctions that retrieve these members, than you define concepts that require these metafunctions... and this all for some types that seem to be the same for every type? Seems like needless over-generalization to me, but maybe we can discuss why its required.

File: include/seqan3/io/detail/chunking.hpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+//
+// template <typename t>
+// using chunk_type_t = typename chunk_type<t>::type;
+// }  // namespace seqan3
+
+namespace seqan3::detail
+{
+
+template <typename iterator_t>
+class chunk_decorator
```
Comment (h-2) [inline include/seqan3/io/detail/chunking.hpp]: whats the point of this?

The general lack of documentation makes this very hard to understand.

File: include/seqan3/core/concept/core.hpp [hunk 6]
```diff
@@ -68,39 +68,39 @@ concept bool implicitly_convertible_to_concept =    static_cast<bool>(ranges::Im
 template <typename t, typename u>
 concept bool explicitly_convertible_to_concept =    static_cast<bool>(ranges::ExplicitlyConvertibleTo<t, u>());
 
-/*!\brief Resolves to `ranges::ConvertibleTo<type1, type2>()`
```
Comment (h-2) [inline include/seqan3/core/concept/core.hpp]: please put whitespace changes into seperate commits

File: include/seqan3/io/detail/direction_iterator_concept.hpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+#pragma once
+
+namespace seqan3::detail
+{
+
+template <typename t>
+concept bool in_direction_iterator_concept<t> = input_iterator_concept<t>;
+
+template <typename t>
+concept bool out_direction_iterator_concept<t> = output_iterator_concept<t> &&
```
Comment (h-2) [inline include/seqan3/io/detail/direction_iterator_concept.hpp]: whats the difference between an out_direction_iterator and an output_iterator?

File: include/seqan3/io/detail/tokenization.hpp [hunk 8]
```diff
@@ -0,0 +1,1175 @@
...
+#include <seqan3/core/meta/associated_types.hpp>
+#include <seqan3/alphabet/concept.hpp>
+#include <seqan3/io/detail/direction_iterator.hpp>
+#include <seqan3/io/detail/stream_iterator.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename t>
+struct Value
```
Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: this is all in CamelCase

Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: and Value<..>::Type needs to be replaced with value_type_t<...>

File: include/seqan3/io/detail/tokenization.hpp [hunk 9]
```diff
@@ -0,0 +1,1238 @@
...
+      output_t && o_ptr)
+    requires input_iterator_concept<std::remove_reference_t<input_t>> &&
+             std::is_base_of_v<chunk_decorator<std::remove_reference_t<input_t>>, std::remove_reference_t<input_t>> &&
+             std::is_pointer_v<std::remove_reference_t<output_t>>
+{
+    // we need the size type of an iterator?
+    using source_size_t = size_type_t<input_t>;
+
+    using std::size;
+    using std::begin;
```
Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: ranges::size and ranges::begin

File: include/seqan3/io/detail/tokenization.hpp [hunk 10]
```diff
@@ -0,0 +1,1238 @@
...
+    requires sized_range_concept<container_t> &&
+             assignable_concept<value_type_t<target_t>&, value_type_t<container_t>>
+{
+    using std::size;
+    auto rng = input_iterator(input);
+    write(std::get<0>(rng), size(input), output);
+}
+
+// ----------------------------------------------------------------------------
+// Function get()
```
Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: think a little more about the name

File: include/seqan3/io/detail/tokenization.hpp [hunk 11]
```diff
@@ -0,0 +1,1238 @@
...
+// ----------------------------------------------------------------------------
+// Function do_get(); Element-wise
+// ----------------------------------------------------------------------------
+
+template <typename input_t, typename in_sent_t,
+          typename output_t,
+          typename stop_predicate_t,
+          typename ignore_predicate_t>
+inline void
+do_get(input_t && i_iter,
```
Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: get_impl

File: include/seqan3/io/detail/tokenization.hpp [hunk 12]
```diff
@@ -0,0 +1,1238 @@
...
+}
+
+// ----------------------------------------------------------------------------
+// Function do_ignore(); Element-wise
+// ----------------------------------------------------------------------------
+
+template <typename input_t, typename in_sent_t,
+          typename stop_predicate_t>
+inline void
+do_ignore(input_t && i_iter,
```
Comment (h-2) [inline include/seqan3/io/detail/tokenization.hpp]: don't create extra ignore/skip interfaces, instead have `get_line(ibeg, iend, std::ignore)`

File: include/seqan3/type_traits/core.hpp [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace seqan3
+{
+
+template <typename ...t>
+    struct common_reference : public ranges::common_reference<t...>
+{};
+
+template <typename ...t>
+using common_reference_t = typename common_reference<t...>::type;
```
Comment (h-2) [inline include/seqan3/type_traits/core.hpp]: use ranges:: instead

**Comment by h-2** [comment]
can we close this in favor of the other PRs?

**Comment by rrahn** [comment]
Will be superseded by multiple smaller PRs

**Pull request closed without merging**